presto-jdbc/src/test/java/com/facebook/presto/jdbc/TestJdbcResultSet.java

```diff
@@ -149,7 +152,8 @@ public void testObjectTypes()
             assertEquals(rs.getObject(15), Timestamp.valueOf(LocalDateTime.of(2018, 2, 13, 13, 14, 15, 123_000_000)));
             assertEquals(rs.getTimestamp(15), Timestamp.valueOf(LocalDateTime.of(2018, 2, 13, 13, 14, 15, 123_000_000)));
 
-            assertEquals(rs.getObject(16), Timestamp.valueOf(LocalDateTime.of(2018, 2, 13, 17, 59, 15, 227_000_000))); // TODO this should represent TIMESTAMP '2018-02-13 13:14:15.227 Europe/Warsaw'
+            assertEquals(rs.getObject(16), ZonedDateTime.of(2018, 2, 13, 13, 14, 15, 227_000_000, ZoneId.of("Europe/Warsaw")));
+            assertThrows(IllegalArgumentException.class, () -> rs.getTimestamp(16));
```

Comment:
I don't think this is the right behavior for JDBC. Generally, the `ResultSet` methods are about the desired type in the Java language, not the type in the database. So if someone asks for `getTimestamp()` on a `timestamp with time zone`, we should give them a `Timestamp`, even if that's different than the `getObject()` value.

Also, this is a backwards incompatible change, and would break anyone doing something like this:

```java 
ResultSet rs = statement.executeQuery("SELECT current_timestamp AS now");
rs.next();
rs.getTimestamp("now");
```
Which seems likely to be a common pattern in real queries.

Are you aware of other databases that properly support `timestamp with time zone`? How do their JDBC drivers handle this?

Reply:
> Are you aware of other databases that properly support timestamp with time zone? How do their JDBC drivers handle this?

+1

Yet even if we see that change like this is inline with what other databases do I think we should not break backward compatibility in JDBC driver behaviour  from one release to the other.
Probably we should have some deprecation policy with period where both new and old behaviour are supported. User is able to pick version of semantics using configuration parameter. Only after some time we remove support for deprecated version. This will give users time to adapt to new situation.

@electrum Is there some process for backward incompatible changes already settled up for Presto project?

Reply:
I see no problem supporting `getTimestamp` for `timestamp w/tz`. In fact, we could consider in the future allowing `getDate` and `getTime` for `timestamp` and `timestamp w/tz` (today, they throw).

For the record, I know two real databases that I know have support for `timestamp w/tz`: Oracle and Teradata. (Postgres supports this data type incorrectly). 

I checked Oracle and I'm not completely satisfied with the result (spaces around `T` added for legibility):
```
Statement statement = connection.createStatement();
ResultSet rs = statement.executeQuery("select TIMESTAMP '1997-01-31 09:26:56.66 +02:00' from dual");
assertTrue(rs.next());

oracle.sql.TIMESTAMPTZ timestamptz = (oracle.sql.TIMESTAMPTZ) rs.getObject(1);
System.out.println(timestamptz.offsetDateTimeValue(connection)); // 1997-01-31 T 09:26:56.660+02:00

System.out.println(rs.getTimestamp(1).toLocalDateTime()); // 1997-01-31 T 08:26:56.660
```
-- I'd expect last line to print "...09:26..." rather than "...08:26...".

(I am not changing anything here until we agree on something.)

Reply:
Also, there is a question how to represent `TIME WITH TIME ZONE`. If we retain the ability to query it with `resultSet.getTime()`, then we can change `resultSet.getObject()` to return `java.time.OffsetTime` (this doesn't represent political time zones, which we want to abandon anyway)

@electrum, two questions to you:
1. should there be some kind of a switch for `rs.getObject()` behavior for `TIMESTAMP W/TZ`? Or it's enough to retain `rs.getTimestamp()`? (the value returned from `rs.getTimestamp()` would be _changed_, the previously returned value is _not_ a local time from the contained value)
2. should this PR address `TIME WITH TIME ZONE` too?

Reply:
1. I think it's fine to change the value. It was never documented and people are likely using either `getTimestamp()` or are simply calling `toString()` on the `getObject()` value.
2. We should fix it, but it doesn't need to be in this PR (though it could be if you want to address it now).